cml: document --single, --reuse and --cloud-startup-script (#2312)
iterative/dvc.org: content/docs/cml/self-hosted-runners.md

| ------------------------ | ------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- |
| `--version`              | Show version number                                                                                           | Boolean                                                                                                        |
| `--labels`               | Comma delimited runner labels                                                                                 | Default: `cml`                                                                                                 |
| `--idle-timeout `        | Time in seconds for the runner to be waiting for jobs before shutting down. 0 waits forever.                  | Default: `300`                                                                                                 |

> | `--name`                 | Name displayed in the repo once registered                                                                    | Default: `cml-` followed by a unique identifier, i.e. `cml-cfwj9rrari`                                         |

Reviewer: I noticed the `Values` column is a bit inconsistent. Sometimes it has a var type e.g. `String` and other times (like here) it specifies the default string instead. Further down there are some defaults in the `Description` too. Would be good to make it consistent one way or another.

Commenting post-merge since it's unrelated to the original intention of the PR.

Author: I'm inclined to think that the best long-term solution would be replacing this section with automatically generated documentation as per my comment on https://github.com/iterative/cml/issues/361.

In the meanwhile, I'll take a look to the table tomorrow and try to unify the current row style sample book.

Author: Resolved through #2321

> | `--single`               | If specified, exit after running a single job.                                                                | Boolean                                                                                                        |

Reviewer: ```suggestion
| `--single`               | Exit after running a single job.                                                                | Boolean                                                                                                        |
```

> | `--reuse`                | Don't launch new runners if there is, at least, another one with the same `--name` or containing `--labels`   | Boolean                                                                                                        |

Reviewer: ```suggestion
| `--reuse`                | Don't launch new runners if one already exists with the same name or labels   | Boolean                                                                                                        |
```

Reviewer: But it's not clear what the condition is. The existing runner must have the same name AND exact same labels? Just name or some of the labels? Would be good to clarify the exact behavior.

Author: 🙏  Thanks, @jorgeorpinel! As I tried to express in the original text, there are two disjunctive conditions: for a runner to be reused, it has to either (1) have the same name or (2) contain, at least, the specified labels.

Would the following modification convey the behavior in a clearer way? If not, suggestions are highly welcome.
```suggestion
| `--reuse`                | Don't launch new runners if one already exists with the same name or containing the specified labels   | Boolean                                                                                                        |
```

Reviewer: Its better, thanks! But I'm still not getting it 100% with regards to labels. So it's either the exact same name or the exact same labels? If so then my suggestion (previous comment, which I just updated) should be OK, I think.

Reviewer: BTW last question on --reuse:

> at least one already exists

Meaning one runner on the same platform (driver)? Or across all drivers? (This may be obvious but I'm not familiar enough with CML yet)

Author: @jorgeorpinel, the issue is that, when reusing a runner by labels we don't require an exact match. _Id est:_ if an existing runner [contains](https://en.wikipedia.org/wiki/Subset) all the given labels among many others, we still consider it a match.

Author: > _Meaning one runner on the same platform (driver)? Or across all drivers? (This may be obvious but I'm not familiar enough with CML yet)_

For users, the so-called drivers are just the name of their continuous integration vendor, like GitHub Actions, GitLab CI/CD, Azure Pipelines, CircleCI... We only support one of these at a time, so «one runner on the same platform» would be the right meaning.

Reviewer: Cools, so maybe
```suggestion
| `--reuse`                | Don't launch a new runner if at an existing one has the same name, or overlapping labels   | Boolean                                                                                                        |
```

Author: > _if **at** an_

Is that `at` a residue from the original sentence?

Author: ```suggestion
| `--reuse`                | Don't launch a new runner if an existing one has the same name, or overlapping labels   | Boolean                                                                                                        |
```

Reviewer: > if at an

Yeah that was a typo sry 😅

> | `--driver`               | If not specified, driver is inferred from environmental variables                                             | Choices: `github`,`gitlab`                                                                                     |

Reviewer: Is "type of driver" the correct way to say this? Or something like "Driver platform" perhaps?

Reviewer: Please ignore this comment on --driver for now. Didn't mean to post this yet. Resolving!

Author: The _driver_ concept is a bit of a stretch and describes a platform-specific continuous integration runner [(like this)](https://github.com/actions/runner) that's wrapped by our code. I like the idea of "driver platform" even if the platform part becomes more insightful than the driver part.

Reviewer: Thanks for the info. It will be useful for a separate review I'm preparing... ⌛

| `--cloud-hdd-size`       | HDD size in GB.                                                                                               | Accepts integer values                                                                                         |
| `--cloud-ssh-private`    | Your private RSA SSH key. If not provided will be generated by the Terraform-provider-Iterative.              | Accepts string                                                                                                 |
| `--cloud-spot`           | Request a spot instance                                                                                       | Boolean                                                                                                        |
| `--cloud-spot-price`     | Spot max price. If not specified it takes current spot bidding pricing.                                       | Default: `-1`                                                                                                  |

> | `--cloud-startup-script` | Run the provided script as part of the instance initialization                                                | String (Base64 encoded shell script)                                                                           |

Reviewer: ```suggestion
| `--cloud-startup-script` | Run the script provided in-line during the instance initialization. Expects a Base64-encoded shell script                                                | String                                                                           |
```

Reviewer: > the instance initialization

Is it only one instance always? Or should it be more general in case of multiple runners e.g. "during runner instance(s) initialization?

Also, would it be important to specify when exactly during initialization this is run?

Reviewer: > shell script

Maybe "Linux shell script"? Or can it be other stuff e.g. BAT (Windows). Also if it's specifically Bash, maybe say "Bash script".

Author: > _Maybe "Linux shell script"?_

For the _connoisseur_, the term [_shell script_](https://en.wikipedia.org/wiki/Shell_script) doesn't leave any doubt about its more-or-less POSIX-compliant nature, but this is just one of my attrocious nitpicking exercises. 😂  I would be tempted to use Bash as you suggested, but the fun fact is that we're using the good old [Bourne shell](https://en.wikipedia.org/wiki/Bourne_shell).

_Linux shell script_ might be the most helpful term for now, but I'll consider migrating to Bash under the hood and updating the documentation. The good old `sh` shell is not precisely famous for its syntactic sugar.

Author: > _Is it only one instance always? Or should it be more general in case of multiple runners e.g. "during runner instance(s) initialization?_

Currently we only support launching a single instance per invocation, so the singular would be pretty safe.

Author: > _Also, would it be important to specify when exactly during initialization this is run?_

The technically inaccurate answer is ["nearly in the middle"](https://github.com/iterative/terraform-provider-iterative/blob/d54e3415ba94fcc7b7c28a9b989b4a21289950dc/iterative/resource_runner.go#L225-L227) between the instance configuration and the runner execution:

1. Install all the base packages and configure the instance operating system if required.
2. Run the custom script and wait until it finishes.
3. Configure and launch the CML self-hosted runner.

Reviewer: OK no need for so much detail in docs involving the script execution order I guess. I edited my suggestion above to include your other answers. Thanks!

Reviewer: Should we link to https://www.base64decode.org/ or similar tool from "Base64-encoded" ? Or perhaps this is explained/linked to from some other CML doc.

Author: As far as I know, the CML documentation doesn't include any user-friendly explanation of Base64. Anyhow, I would prefer not to link to any specific on-line tool.

If you consider it neccessary, we can recommend the use of the `base64` command-line tool and link to something neutral, like the pertaining Wikipedia article, but I presume that users searching for this feature will be somewhat familiar with this encoding.

Reviewer: > CML documentation doesn't include any user-friendly explanation of Base64

Might be something to make an issue to improve on. Probably not a good idea to assume users will know what it even means.

Author: Agreed! In fact, we probably should start by reviewing if the special character issues we're solving with Base64 are worth the impact from the user experience standpoint.

Note to myself: write a nice issue about this concern on the CML repository.

Reviewer: Up to you guys @0x2b3bfa0 @DavidGOrtega unfort. currently the docs team has to focus mostly on DVC so can't follow up on this, but hopefully soon we will be able to expand a little and help you more closely 🙂

Author: Awesome! Moved to https://github.com/iterative/terraform-provider-iterative/issues/91 in the meanwhile
